### PR TITLE
Story/quad 46

### PR DESCRIPTION
**Pull request by suprajkolluri**
Refactored the code and fixed bugs pertaining to rest controllers

### CODE REVIEW

File: Quadriga/src/main/java/edu/asu/spring/quadriga/aspects/RestAccessAspect.java [hunk 1]
```diff
@@ -11,92 +13,101 @@
...
-		if(!haveAccess)
-		{
-			return restAccessException.accessErrorMsg(400,"You don't have permission");
+
+		if (!haveAccess) {
+			throw new RestException(403);
+			// return new ResponseEntity<String>("<error>You don't have
+			// permission</error>", HttpStatus.UNAUTHORIZED);
+			// return restAccessException.accessErrorMsg(400, "You don't have
+			// permission");
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/aspects/RestAccessAspect.java:107]: delete, not just comment out (same above)

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/DictionaryRestController.java [hunk 2]
```diff
@@ -74,391 +76,348 @@
...
+            return new ResponseEntity<String>(writer.toString(), HttpStatus.OK);
+        } catch (ResourceNotFoundException e) {
+            logger.error("Exception:", e);
+            throw new RestException(404);
+        } catch (ParseErrorException e) {
+            logger.error("Exception:", e);
+            throw new RestException(404);
+        } catch (MethodInvocationException e) {
+            logger.error("Exception:", e);
+            throw new RestException(403);
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/DictionaryRestController.java:147]: wrap the original exception in the new one and remove the logger (same in the ohter classes).

and do that everywhere

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java [hunk 3]
```diff
@@ -73,425 +75,418 @@
...
+    public ResponseEntity<String> addConceptsToConceptColleciton(
+            @PathVariable("conceptCollectionID") String conceptCollectionId, HttpServletRequest request,
+            HttpServletResponse response, @RequestBody String xml, @RequestHeader("Accept") String accept,
+            Principal principal) throws QuadrigaException, ParserConfigurationException, SAXException, IOException,
+                    JAXBException, QuadrigaAccessException, QuadrigaStorageException, RestException {
+        IUser user = userManager.getUser(principal.getName());
+        if (xml.equals("")) {
+            String errorMsg = restMessage.getErrorMsg("Please provide XML in body of the post request.", request);
+            return new ResponseEntity<String>(errorMsg, HttpStatus.NOT_FOUND);
+        } else {
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java:208]: loose this else. It's unnecessary. the return statement befreo will make sure, the code after the if will never be reached.

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java [hunk 4]
```diff
@@ -73,425 +75,418 @@
...
+            try {
+                JAXBContext context = JAXBContext.newInstance(QuadrigaConceptReply.class);
+                Unmarshaller unmarshaller = context.createUnmarshaller();
+                unmarshaller.setEventHandler(new javax.xml.bind.helpers.DefaultValidationEventHandler());
+                InputStream is = new ByteArrayInputStream(xml.getBytes());
+                response1 = unmarshaller.unmarshal(new StreamSource(is), QuadrigaConceptReply.class);
+            } catch (Exception e) {
+                logger.error("Error in unmarshalling", e);
+                String errorMsg = restMessage.getErrorMsg("Error in unmarshalling", request);
+                return new ResponseEntity<String>(errorMsg, HttpStatus.FORBIDDEN);
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java:221]: why do you set error "forbidden" if it's a generic error message?

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java [hunk 5]
```diff
@@ -73,425 +75,418 @@
...
+            }
+            QuadrigaConceptReply qReply = response1.getValue();
+            ConceptList c1 = qReply.getConceptList();
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java:224]: c1 is a bad variable name

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java [hunk 6]
```diff
@@ -73,425 +75,418 @@
...
+            List<Concept> conceptList = c1.getConcepts();
+
+            Iterator<Concept> I = conceptList.iterator();
+
+            while (I.hasNext()) {
+                Concept c = I.next();
+                logger.debug("arg Name :" + c.getName().trim());
+                logger.debug("arg Pos :" + c.getPos().trim());
+                logger.debug("arg URI :" + c.getUri().trim());
+                logger.debug("arg descrtiption :" + c.getDescription().trim());
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java:234]: loose all these debug messages. if you really want to print the concept as debug msg, override the toString method and use that one instead. (one line of code)

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java [hunk 7]
```diff
@@ -73,425 +75,418 @@
...
+        if (!checkWSSecurity.checkIsWorkspaceExists(workspaceId)) {
+            String errorMsg = restMessage.getErrorMsg("Workspace ID : " + workspaceId + " doesn't exist", request);
+            return new ResponseEntity<String>(errorMsg, HttpStatus.NOT_FOUND);
+        }
+
+        IConceptCollection collection = conceptCollectionFactory.createConceptCollectionObject();
+
+        if (ccName == null || ccName.isEmpty()) {
+            String errorMsg = restMessage.getErrorMsg("Please provide concept collection name", request);
+            return new ResponseEntity<String>(errorMsg, HttpStatus.NOT_FOUND);
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java:367]: should not be "not found" but another more appropriate error status.

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java [hunk 8]
```diff
@@ -73,425 +75,418 @@
...
+        }
+        if (desc == null || desc.isEmpty()) {
+            String errorMsg = restMessage.getErrorMsg("Please provide concept collection description", request);
+            return new ResponseEntity<String>(errorMsg, HttpStatus.NOT_FOUND);
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java:371]: same here

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java [hunk 9]
```diff
@@ -73,425 +75,418 @@
...
+            logger.error("Error in unmarshalling", e);
+            String errorMsg = restMessage.getErrorMsg("Error in unmarshalling", request);
+            return new ResponseEntity<String>(errorMsg, HttpStatus.FORBIDDEN);
+        }
+        if (response1 == null) {
+            String errorMsg = restMessage.getErrorMsg("Concepts XML is not valid", request);
+            return new ResponseEntity<String>(errorMsg, HttpStatus.NOT_FOUND);
+        }
+        QuadrigaConceptReply qReply = response1.getValue();
+        ConceptList c1 = qReply.getConceptList();
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java:391]: bad variable name

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java [hunk 10]
```diff
@@ -73,425 +75,418 @@
...
+        List<Concept> conceptList = c1.getConcepts();
+        if (conceptList.size() < 1) {
+            String errorMsg = restMessage.getErrorMsg("Concepts XML is not valid", request);
+            return new ResponseEntity<String>(errorMsg, HttpStatus.NOT_FOUND);
+        }
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java:396]: httpstatus wrong

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java [hunk 11]
```diff
@@ -73,425 +75,418 @@
...
+
+        collection.setDescription(desc);
+        collection.setOwner(user);
+        collection.setConceptCollectionName(ccName);
+
+        conceptControllerManager.addConceptCollection(collection);
+        String ccId = conceptControllerManager.getConceptCollectionId(ccName);
+
+        Iterator<Concept> I = conceptList.iterator();
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java:405]: variable name

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java [hunk 12]
```diff
@@ -73,425 +75,418 @@
...
+
+        while (I.hasNext()) {
+            Concept c = I.next();
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java:408]: variable name

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java [hunk 13]
```diff
@@ -73,425 +75,418 @@
...
+            throw new RestException(500);
+        } catch (QuadrigaAccessException e) {
+            logger.error("Exception:", e);
+            throw new RestException(403);
+        } catch (Exception e) {
+            logger.error("Exception:", e);
+            throw new RestException(403);
+        }
+
+    }
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java:490]: please check the whole file for my comments (eg. all variable names, logger output, etc).

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/DictionaryRestController.java [hunk 14]
```diff
@@ -74,391 +76,348 @@
...
+                    return new ResponseEntity<String>(errorMsg, httpHeaders, HttpStatus.INTERNAL_SERVER_ERROR);
+                }
+
+            }
+
+            HttpHeaders httpHeaders = new HttpHeaders();
+            httpHeaders.setContentType(MediaType.valueOf(accept));
+            return new ResponseEntity<String>("Success", httpHeaders, HttpStatus.OK);
+        }
+    }
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/DictionaryRestController.java:421]: please check this file for the comments I made to the other file

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/NetworkRestController.java [hunk 15]
```diff
@@ -48,661 +50,636 @@
...
+        String networkId = "";
+        String workspaceid = request.getParameter("workspaceid");
+
+        if (workspaceid == null || workspaceid.isEmpty()) {
+            String errorMsg = errorMessageRest.getErrorMsg("Please provide correct workspace id.");
+            return new ResponseEntity<String>(errorMsg, HttpStatus.NOT_FOUND);
+        }
+        String projectid = networkManager.getProjectIdForWorkspaceId(workspaceid);
+        if (projectid == null || projectid.isEmpty()) {
+            String errorMsg = errorMessageRest.getErrorMsg("Please provide correct workspace id.");
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/NetworkRestController.java:122]: This is not the correct error message here. Should tell the user that the project id couldn't be found for the given workspace id (this should never happen).

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/NetworkRestController.java [hunk 16]
```diff
@@ -48,661 +50,636 @@
...
+            return new ResponseEntity<String>(errorMsg, HttpStatus.NOT_FOUND);
+        }
+
+        if (networkName == null || networkName.isEmpty()) {
+            String errorMsg = errorMessageRest.getErrorMsg("Please provide network name as a part of post parameters");
+            return new ResponseEntity<String>(errorMsg, HttpStatus.NOT_FOUND);
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/NetworkRestController.java:128]: check httpstatus in class

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/NetworkRestController.java [hunk 17]
```diff
@@ -48,661 +50,636 @@
...
+        }
+
+        xml = xml.trim();
+        if (xml.isEmpty()) {
+            String errorMsg = errorMessageRest.getErrorMsg("Please provide XML in body of the post request.");
+            return new ResponseEntity<String>(errorMsg, HttpStatus.NOT_FOUND);
+
+        } else {
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/NetworkRestController.java:136]: you don't need this else; if it's going into the if, it'll return in any case

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/NetworkRestController.java [hunk 18]
```diff
@@ -48,661 +50,636 @@
...
+            String res = networkManager.storeXMLQStore(xml);
+            if (res.equals("")) {
+                String errorMsg = "Please provide correct XML in body of the post request. Qstore system is not accepting ur XML";
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/NetworkRestController.java:139]: please correct spelling

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/NetworkRestController.java [hunk 19]
```diff
@@ -48,661 +50,636 @@
...
+                return new ResponseEntity<String>(errorMsg, HttpStatus.NOT_FOUND);
+            }
+
+            networkId = networkManager.storeNetworkDetails(res, user, networkName, workspaceid,
+                    INetworkManager.NEWNETWORK, networkId, INetworkManager.VERSION_ZERO);
+
+            if (networkId.endsWith(INetworkManager.DSPACEERROR)) {
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/NetworkRestController.java:146]: I think we don't need this block anymore; please double-check and delete.

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/NetworkRestController.java [hunk 20]
```diff
@@ -48,661 +50,636 @@
...
+            context.put("networkid", networkId);
+            List<INetworkAnnotation> networkAnnoList = editingNetworkAnnoManager
+                    .getAllAnnotationOfNetwork(user.getUserName(), networkId);
+            context.put("networkAnnoList", networkAnnoList);
+            StringWriter writer = new StringWriter();
+            template.merge(context, writer);
+
+            HttpHeaders httpHeaders = new HttpHeaders();
+            httpHeaders.setContentType(MediaType.APPLICATION_XML);
+            return new ResponseEntity<String>(writer.toString(), httpHeaders, HttpStatus.OK);
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/NetworkRestController.java:204]: move everything out of this try block and only put try catch around the code that actually throws the exception (e.g. building the context can probably be done before the try).

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/NetworkRestController.java [hunk 21]
```diff
@@ -48,661 +50,636 @@
...
+            String errorMsg = errorMessageRest.getErrorMsg("The Network doesn't have status : REJECTED");
+            return new ResponseEntity<String>(errorMsg, HttpStatus.NOT_FOUND);
+        } else {
+
+            xml = xml.trim();
+            if (xml.isEmpty()) {
+                String errorMsg = errorMessageRest.getErrorMsg("Please provide XML in body of the post request.");
+                return new ResponseEntity<String>(errorMsg, HttpStatus.NOT_FOUND);
+
+            } else {
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/NetworkRestController.java:648]: loose the else

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java (hunk 4 above)
Comment (suprajkolluri) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java:221]: In the article(http://www.restapitutorial.com/httpstatuscodes.html ) I read about Error 403 as -  
403 Forbidden
The server understood the request, but is refusing to fulfill it. 

Also the second comment here - http://stackoverflow.com/questions/3290182/rest-http-status-codes-for-failed-validation-or-invalid-duplicate
Failed validation: 403 Forbidden ("The server understood the request, but is refusing to fulfill it"). Contrary to popular opinion, RFC2616 doesn't say "403 is only intended for failed authentication", but "403: I know what you want, but I won't do that". That condition may or may not be due to authentication.

Therefore I thought the error code was appropriate. Please let me know if you want me to change it.

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/NetworkRestController.java [hunk 22]
```diff
@@ -48,661 +50,636 @@
...
+                }
+
+                // networkManager.archiveNetwork(networkId);
+                editorManager.updateNetworkStatus(networkId, INetworkStatus.PENDING);
+                int latestVersion = networkManager.getLatestVersionOfNetwork(networkId) + 1;
+                networkId = networkManager.storeNetworkDetails(res, user, networkName,
+                        network.getNetworkWorkspace().getWorkspace().getWorkspaceId(), INetworkManager.UPDATENETWORK,
+                        networkId, latestVersion);
+
+                if (networkId.endsWith(INetworkManager.DSPACEERROR)) {
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/NetworkRestController.java:672]: you can probably delete this if block

File: Quadriga/src/main/java/edu/asu/spring/quadriga/service/impl/workspace/ListWSManager.java [hunk 23]
```diff
@@ -263,8 +263,10 @@ public String getWorkspaceName(String workspaceId) throws QuadrigaStorageExcepti
                 }
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/service/impl/workspace/ListWSManager.java:263]: in line 250, negate the condition and move the code from the else block into the if block.

in line 244 initialize the removeList and then loose line 253 to 254 and 259 to 261

in line 246, turn the for loop into a for (IWorkspaceNetwork workspaceNetwork : workspaceNetworkList) loop

File: Quadriga/src/main/resources/locale/errormessages.properties [hunk 24]
```diff
@@ -1,6 +1,7 @@
 error_message_401 = Unauthorized Access. Please check login credentials.
-error_message_403 = Not authorized to access this page.
+error_message_403 = Not authorized to access this page or the server is unable to full fill the request.
```
Comment (jdamerow) [inline Quadriga/src/main/resources/locale/errormessages.properties:2]: You're not authorized to access this page or the server is unable to fulfill the request. 

When is that thrown? This are two different http status, and should not be put into one answer.

File: Quadriga/src/main/resources/locale/errormessages.properties [hunk 25]
```diff
@@ -1,6 +1,7 @@
 error_message_401 = Unauthorized Access. Please check login credentials.
-error_message_403 = Not authorized to access this page.
+error_message_403 = Not authorized to access this page or the server is unable to full fill the request.
 error_message_404 = Resource not found
 error_message_405 = Database is down!! People are working on it
```
Comment (jdamerow) [inline Quadriga/src/main/resources/locale/errormessages.properties:4]: The database is down. If this error persist please report the issue.

**Comment by jdamerow** [comment]
see comments

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/DictionaryRestController.java (hunk 2 above)
Comment (suprajkolluri) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/DictionaryRestController.java:147]: Shouldn't we log error messages?

**Pull request closed without merging**